File: baekjoon/bruteforce/9019/Main.java
/**
 * time over
 **/
import java.util.*;

class Main {
    static final String ch = "DSLR";
    static int a,b;
    
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);    
        int testCase = sc.nextInt();
        for(int tc=0; tc<testCase; tc++) {
            a = sc.nextInt();
            b = sc.nextInt();
            
            Queue<Integer> q = new LinkedList<Integer>();
            
            int[][] parent = new int[10001][2];
            for(int i=0; i<10001; i++) {
                Arrays.fill(parent[i], -1);
            }
            
            parent[a][0] = a;
            parent[a][1] = -1;
            q.add(a);
            
            while(!q.isEmpty()) {
                int here = q.peek();
                q.remove();
                
                if(here == b) break;
                for(int i=0; i<4; i++) {
                    int there = operate(here, i);
                    if(parent[there][0] == -1) {
                        parent[there][0] = here;
                        parent[there][1] = i;
                        q.add(there);
                    }
                }
            }
            
            int n;
            List<Character> ret = new ArrayList<Character>();
            while(a != b) {
                n = parent[b][1];
                ret.add(ch.charAt(n));
                b = parent[b][0];
            }
            Collections.reverse(ret);
            for(int i=0; i<ret.size(); i++) {
                System.out.print(ret.get(i));
            }
        }
    }
    
    public static int operate(int n, int op) {
        switch(op) {
            case 0:
                n = (n*2)%10000;
                break;
            case 1:
                if(n == 0) n = 9999; 
                else n--;
                break;
            case 2:
                n = (n*10)%10000 + (n/1000);
                break;
            case 3:
                n = n/10 + (n%10)*1000;
                break;
            default :
                break;
        }
        return n;
    }
}

File: baekjoon/bruteforce/9019/main.cpp
#include <iostream>
#include <cstdio>
#include <queue>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;
const int oo = 0x3f3f3f3f;
const string ch = "DSLR";
int operate(int n, int op);
int main()
{
    int testCase;
    scanf("%d", &testCase);
    while(testCase--) {
        int a,b;
        scanf("%d %d", &a, &b);
        queue<int> q;
        q.push(a);
        vector<pair<int,int> > parent = vector<pair<int, int> >(10001, make_pair(-1,-1));
        parent[a].first = a;
        while(!q.empty()) {
            int here = q.front();
            q.pop();
            if(here == b) break;
            for(int i=0; i<4; i++) {
                int there = operate(here, i);
                if(parent[there].first == -1) {
                    parent[there].first = here;
                    parent[there].second = i;
                    q.push(there);
                }
            }
        }
        int n;
        string ret;
        while(b != a) {
            n = parent[b].second;
            ret.push_back(ch[n]);
            b = parent[b].first;
        }
        reverse(ret.begin(), ret.end());
        cout << ret << endl;
    }
    return 0;
}


int operate(int n, int op)
{
    switch(op) {
        case 0:
            n = (n*2)%10000;
            break;
        case 1:
            if(n == 0) n = 9999; 
            else n--;
            break;
        case 2:
            n = (n*10)%10000 + (n/1000);
            break;
        case 3:
            n = n/10 + (n%10)*1000;
            break;
        default :
            break;
    }
    return n;
}